Full Refactor & 'Inclined Planes' ReWrite

acbuynak opened the pull request:
Refactor of code.

Full details of refactor documented in Issue #4.

Schellenberg3 committed:
completes #4

acbuynak commented:
@Schellenberg3 
Reviewing the git log, I think I figured out what was 'weird' about the commit you made.

I think you somehow either merged (or manually copied in?) my earlier commits and pushed them in as a single commit. 
The other more likely case is you somehow pulled down my commits in the `refactor` branch but weren't writing code at the `head` of the branch (?). So when you pushed it probably asked you to `force-push` and it overwrote my commits. 

* `force-push` on shared repo's is rarely a good idea. The only time I ever actively use it is for rebasing commits.
* Please breakup your changes into different commits so we can trace what changed easier, but I get that's less realistic for a refactor.
*  Looking at the change log at this point is not realistic since entire files just were overwritten instead of modified. Similarly, I can't leave review comments directly on the code bc of this since so many things moved/changed at once.
 
Thanks for the code. I don't mean to chastise, just trying to help explain my side since I know we've been trying to learn git better.

acbuynak commented on scripts/transformations.py:
```diff
@@ -20,130 +20,137 @@
-
-##########################
-
+from typing import Union
+from typing import List
```
These imports are not used. Further they are very close to existing reserved names of `list`.
Why do these need included?

Schellenberg3 replied:
They are for the type assistance / type hints under after function - `List` is intentionally close to `list`. Though I don't know *exactly* how the type assists work this is the proper way to signify a list. (`Union` signifies that a function can accept multiple input type for one argument).

They aren't and shouldn't be used in the code itself per se but are needed for the type assistance. [See here](https://mypy.readthedocs.io/en/stable/python2.html)

The short of it is that they should stay but I'm glad you pointed out this potential confusion!

acbuynak replied:
Not a huge fan of including these imports just for the type assistance as this is a not a standard install Python module and would introduce a bug that users would consistently get when setting up environment.

We could including `typing` in the `ros-setup` bash script to ensure inclusion. And also including `typing` in the package`__init.py__` .

So, is this important enough you want to keep it?

Schellenberg3 replied:
Resolved by adding requirements.txt

acbuynak commented on scripts/shapes.py:
```diff
@@ -1,23 +1,49 @@
+from typing import List
```
These imports are not used. Further they are very close to existing reserved names of list.
Why do these need included?

(same commend for import in other scripts)

Schellenberg3 replied:
[See above.](https://github.com/osu-capstone-afrl/photogrammetry-capture-motion/pull/6#discussion_r602890426)

acbuynak commented on scripts/path_plans.py:
```diff
@@ -1,250 +1,140 @@
-    vec = vec / np.sqrt(np.sum(vec**2))
-    return vec.tolist()
+from shapes import Ring
+from typing import List
+from typing import Tuple
+from typing import Dict
```
These imports are not used. Further they are very close to existing reserved names of list.
Why do these need included?

(same comment for similar imports in other files)

Schellenberg3 replied:
[See above](https://github.com/osu-capstone-afrl/photogrammetry-capture-motion/pull/6#discussion_r602890426)

acbuynak commented on scripts/path_plans.py:
```diff
@@ -260,80 +150,217 @@ def _rotationMatrix_orient_z_axis(self, local_point):
         return rot_matrix
 
 
+class NewInclinedPlane(DetectedObject):
```
Since you have deleted the original inclined planes code, I would propose not calling this the 'new' method. I'd just stick with the name `InclinedPlane` as before.

Schellenberg3 replied:
Agreed - forgot to change it back after removing the old code 😛

acbuynak replied:
Cool. Can you push a commit and resolve when done?

acbuynak commented on scripts/path_plans.py:
```diff
@@ -260,80 +150,217 @@ def _rotationMatrix_orient_z_axis(self, local_point):
+        tf_yz2 = self.tf.get_transformation(rot_yz2, plane_yz2_offset)
+
+        no_rotation = np.identity(3)
+        tf_final_offset = self.tf.get_transformation(no_rotation, [0, 0, -1*self._offset])
```
Creating Offset TF, See next comment as to usage.

acbuynak commented on scripts/path_plans.py:
```diff
@@ -260,80 +150,217 @@ def _rotationMatrix_orient_z_axis(self, local_point):
+                # place around part and apply offset
+                tf_around_part = np.matmul(self._locator_tf, tf_around_origin)
+
+                # post to move relative to each frame instead of the world frame
+                tf_with_offset = np.matmul(tf_around_part, tf_final_offset)
```
![image](https://user-images.githubusercontent.com/44533530/112757277-15a73980-8fb7-11eb-96a9-d0a9789a1edd.png)

Noting here the effects of applying offset along the camera frame's z-axis. The curved like shape is unsurprising given the technique we are using. I think it is acceptable as is.

If desired to change back to the entire plane moving, would need to change so offset is based on the fixed frame... but then you need to account for rotation as well. So probably not worth the hassle.

Schellenberg3 replied:
Agreed. I think the only issue here is that it might be unexpected for an user unfamiliar with the code so if we leave it as it (which it sounds like we are) I think we should make sure it is documented.

And yeah, the other way is doable and would require us to assign each point a temporary frame where Z is normal to the plane, then offset, then assign the correct rotation matrix. But I don't feel like writing the code to do all that right now.

acbuynak replied:
I'm comfortable with no change moving forward. We'll need to document well. I might suggest doing such documentation in the Repo Wiki.

acbuynak commented on scripts/path_plans.py:
```diff
@@ -260,80 +150,217 @@ def _rotationMatrix_orient_z_axis(self, local_point):
+        @param size:        list[x,y,z] dimensions of presented part
+        @param locator:     center of presented part in world frame coordinates
+        @param orientation: rotation of presented part in world frame coordinates
+        @param count:       Tuple (X,Z) with the count of points along each direction of the plane
```
Took me a second to realize what this tuple was composed of in regards to which dimension was effected by which variable, but upon checking documentation here the (X,Z) fixed my concerns. Given the assumption that the concepts of TF's is understood by future developers, I think this is sufficient.

Schellenberg3 replied:
You have a good point. I'll think about how this could be clarified by explicitly saying the length and height of the planes.

Schellenberg3 commented:
No worries, the feedback is appreciate! 

I wonder if there are any good-practices for commits when refractoring? Per [this](https://softwareengineering.stackexchange.com/questions/104407/how-to-commit-in-progress-refactoring) it sounds like the goal should have been a set of incremental changes that maintain the codes functionality but change the internals just a bit. What I did was more akin to a full re-write.

But in the end, absolutely need to learn better git! I'll reach out if I hit run into this In the future or I'll create a new remote branch that can be merged in when we're ready.

Schellenberg3 commented:
Leaving the merge action to the pro @acbuynak

acbuynak commented on scripts/path_plans.py:
```diff
@@ -260,80 +150,217 @@ def _rotationMatrix_orient_z_axis(self, local_point):
+        @param count:       Tuple (X,Z) with the count of points along each direction of the plane
+        @param clearance:   Distance between each plane and the part before the planes are inclined
+        @param plane_scale: Tuple (X,Z) describing the how the planes dimensions are scaled relative to the part's
+        @param slope:       Multiplier from 0 (no tilt) to 1 (maximum tilt) to adjust the incline of each plane
```
Scalar Multiplier makes sense. What is maximum tilt? And does that change given different sized objects?

Schellenberg3 replied:
Maximum tilt means the plane is coincident with the bounding box. Will clarify if the docstring.

acbuynak commented on scripts/transformations.py:
```diff
@@ -20,130 +20,137 @@
+
+            path_poses.append([x, y, z, qx, qy, qz, qw])
+
+        return path_poses
+
+
+# todo Remove or refactor this code. For @ACBuynak
 def main():
-  print("-- DEMO ONLY --\nACTUAL TOOLS STORED IN 'transformations' CLASS\n")
-
-  ## Imports
-  import numpy as np
+  print "-- DEMO ONLY --\nACTUAL TOOLS STORED IN 'transformations' CLASS\n"
 
-  ## Instantiate Class
-  tf = transformations()
+  tf = Transformations()
```
Just pointing this out here as an example.. 
Throughout this commit you overwrite code that already exists with the exact same thing, these sort of non-changes should be avoided as it makes it more difficult to trace through the history of the code and see what was worked on or where new bugs could have been introduced.

Ideally, you can avoid this through just using a git client to build your commit and include/exclude the real changes you want to push.

Schellenberg3 replied:
Are you trying to push GitKraken again? 😛  Maybe I should start using that. Maybe it'd help me understand **just** how much I've messed up my commits haha

acbuynak replied:
Maybe. Depending on how you've been using PyCharm, you might be getting the same benefit already.

acbuynak reviewed (commented):
Overall, I think this refactor looks great and really helps clean up the clarity of the code. Checkout some of the comments I've left in this review and we can discuss if anything further should be changed prior to merge.

Testing with the RVIZ PoseArray, it appears everything is correctly orienting.
![image](https://user-images.githubusercontent.com/44533530/112757840-60c24c00-8fb9-11eb-97c5-f0d489f22567.png)
![image](https://user-images.githubusercontent.com/44533530/112758045-4d63b080-8fba-11eb-9ff2-ccb7eaeb77af.png)

acbuynak commented:
Sounds good, when you make changes.. just push them up to the same `melodic-refactor` branch and they'll be auto added into the PR. (The PR is just pulling all commits in branch, not specific commits).

acbuynak commented on scripts/path_plans.py:
```diff
@@ -1,250 +1,140 @@
+        msg = {'Point': [x, y, z], 'Quaternion': [x, y, z, w]}
 
-    def _rotationMatrix_orient_z_axis(self, local_point):
+        @param path_poses: List of poses as [x, y, z, qx, qy, qz, qw]
+        @return: list of poses formatted as a dictionary
+        """
+        path_messages = []
+        for pose in path_poses:
+            msg = {
+                'Point': pose[:3],
+                'Quaternion': pose[3:]
+            }
+            path_messages.append(msg.copy())
+
+        return path_messages
```
Just caught this when testing in ROS code. You cannot  assemble a ROS message like this. 
You need to build ROS messages by importing the object.

acbuynak replied:
I now realize this wasn't supposed to be a ROS message, but an internal formatting thing we were doing previously.
Let's go directly to a ROS message. I'm working on a commit that will do this.

acbuynak committed:
Standardize to ROS geom.Pose Format

acbuynak committed:
Fix Variables calls to match Refactor

acbuynak committed:
Fix: Removed New from NewInclinedPlane name

acbuynak committed:
Misc Changes allowing Motion Demo to function

acbuynak commented:
After above commits, successfully ran simulation to move robot through both path planning types for the same sized part.

Side note, future work will need done (outside of code) to smartly place the part in space such that the robot can 'reach' the spot. My demo showed this could be an issue at times with the limited reach of the GP7 on larger parts.

Schellenberg3 committed:
add requierments.txt

Adds requirements.txt file with typing requirement and updates the instructions on the readme to instll it.

Schellenberg3 committed:
add numpy to requirements.txt

Schellenberg3 committed:
remove unused comment line
